File: Embarcadero/Berlin/CPP/NotesResource/NoteTypesU.cpp
//---------------------------------------------------------------------------

// This software is Copyright (c) 2015 Embarcadero Technologies, Inc.
// You may only use this software if you are an authorized licensee
// of an Embarcadero developer tools product.
// This software is considered a Redistributable as defined under
// the software license agreement that comes with the Embarcadero Products
// and is subject to that software license agreement.

//---------------------------------------------------------------------------

#pragma hdrstop

#include "NoteTypesU.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)

String TNoteJSON::TNames::Title = "title";
String TNoteJSON::TNames::Content = "content";
String TNoteJSON::TNames::Id = "id";

// ---------------------------------------------------------------------------
__fastcall TNote::TNote(const String ATitle, const String AContent):
	FTitle(ATitle), FContent(AContent)
{

}
// ---------------------------------------------------------------------------
__fastcall TNote::TNote(const String ATitle, const String AContent,
	const String &AID) : FID(AID), FTitle(ATitle), FContent(AContent)
{

}
// ---------------------------------------------------------------------------
TNote* __fastcall TNoteJSON::JSONToNote(const TJSONValue * AJSON)
{
	TJSONObject * obj = static_cast<TJSONObject*>(const_cast<TJSONValue*>(AJSON));
        TJSONPair * content = obj->Get(TNames::Content);
        TJSONPair * id = obj->Get(TNames::Id);
	TNote * note = new TNote(
		obj->Get(TNames::Title)->JsonValue->Value(),
		content ? content->JsonValue->Value() : String(""),
		id ? id->JsonValue->Value() : String(""));
	return note;
}
 // ---------------------------------------------------------------------------
std::vector<TNote*> __fastcall TNoteJSON::JSONToNotes(const TJSONArray * AJSON)
{
	std::vector<TNote*> notes;
	for (int i = 0; i < const_cast<TJSONArray*>(AJSON)->Count; i++) {
		notes.push_back(TNoteJSON::JSONToNote(const_cast<TJSONArray*>(AJSON)->Items[i]));
	}
	return notes;
}
// ---------------------------------------------------------------------------
TJSONObject* __fastcall TNoteJSON::NoteToJSON(const TNote *ANote)
{
	TJSONObject * jsonObj = new TJSONObject();
	jsonObj->AddPair(TNames::Title, ANote->Title);
	jsonObj->AddPair(TNames::Content, ANote->Content);
	jsonObj->AddPair(TNames::Id, ANote->ID);
	return jsonObj;
}
// ---------------------------------------------------------------------------
TJSONArray * __fastcall TNoteJSON::NotesToJSON(std::vector<TNote*> *ANotes)
{
	TJSONArray * jsonArray = new TJSONArray();
	for(std::vector<TNote*>::iterator it = ANotes->begin(); it != ANotes->end(); it++)
	{
		jsonArray->Add(TNoteJSON::NoteToJSON(*it));
	}
	return jsonArray;
}
// ---------------------------------------------------------------------------

File: Embarcadero/Rio/Object Pascal/NotesResource/NoteTypesU.pas
//---------------------------------------------------------------------------

// This software is Copyright (c) 2015 Embarcadero Technologies, Inc.
// You may only use this software if you are an authorized licensee
// of an Embarcadero developer tools product.
// This software is considered a Redistributable as defined under
// the software license agreement that comes with the Embarcadero Products
// and is subject to that software license agreement.

//---------------------------------------------------------------------------

unit NoteTypesU;

interface

uses System.JSON;

type
  TNote = record
  private
    FTitle: string;
    FContent: string;
    FID: string;
  public
    constructor Create(const ATitle, AContent: string);
    constructor CreateWithID(const ATitle, AContent: string; const AID: string);
    property Title: string read FTitle write FTitle;
    property Content: string read FContent write FContent;
    property ID: string read FID write FID;
  end;

  TNoteJSON = class
  public type
    TNames = record
    public const
      Title = 'title';
      Content = 'content';
      Id = 'id';
    end;
  public
    class function JSONToNote(const AJSON: TJSONValue): TNote; static;
    class function JSONToNotes(const AJSON: TJSONArray): TArray<TNote>; static;
    class procedure NotesToJSON(const ANotes: TArray<TNote>; const AJSON: TJSONArray); static;
    class function NoteToJSON(const ANote: TNote): TJSONObject; static;
  end;

implementation

uses System.Generics.Collections;

{ TNote }

constructor TNote.Create(const ATitle, AContent: string);
begin
  FContent := AContent;
  FTitle := ATitle;
end;

class function TNoteJSON.JSONToNote(const AJSON: TJSONValue): TNote;
begin
  Result := TNote.CreateWithID(
    AJSON.GetValue<string>(TNames.Title, ''),
    AJSON.GetValue<string>(TNames.Content, ''),
    AJSON.GetValue<string>(TNames.ID, ''));
end;

class function TNoteJSON.NoteToJSON(const ANote: TNote): TJSONObject; // Caller must free result
begin
  Result := TJSONObject.Create;
  Result.AddPair(TNames.Title, ANote.Title);
  Result.AddPair(TNames.Content, ANote.Content);
  Result.AddPair(TNames.ID, ANote.ID);
end;

class function TNoteJSON.JSONToNotes(const AJSON: TJSONArray): TArray<TNote>;
var
  LValue: TJSONValue;
  LList: TList<TNote>;
begin
  LList := TList<TNote>.Create;
  try
    for LValue in AJSON do
      LList.Add(TNoteJSON.JSONToNote(LValue));
    Result := LList.ToArray;
  finally
    LList.Free;
  end;
end;

class procedure TNoteJSON.NotesToJSON(const ANotes: TArray<TNote>;
  const AJSON: TJSONArray);
var
  LNote: TNote;
  LJSON: TJSONObject;
begin
  for LNote in ANotes do
  begin
    LJSON := NoteToJSON(LNote);
    AJSON.Add(LJSON);
  end;
end;

constructor TNote.CreateWithID(const ATitle, AContent, AID: string);
begin
  Create(ATitle, AContent);
  FID := AID;
end;

end.
